### PR TITLE
SALTO-2683/invalid permission types

### PR DESCRIPTION
**Pull request by guyhamo**
added a filter to deploy permission schemes without permissions that are invalid.
also added a change validator to warn the user that those permissions are going to be excluded from the deploy.

---

_Additional context for reviewer_
the added test for permission scheme filter is failing because of a bad nested expect.... probably needs a better check

---
_Release Notes_: 
_jira adapter_:
* exclude permissions that are not part of the environment when deploying permission schemes

---
_User Notifications_:

### CODE REVIEW

File: packages/jira-adapter/src/change_validators/permission_type.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import { PERMISSION_SCHEME_TYPE_NAME, PERMISSIONS } from '../constants'
+
+const { awu } = collections.asynciterable
+
+const log = logger(module)
+
+export const getAllowedPermissionTypes = async (
+  elementSource: ReadOnlyElementsSource,
+): Promise<string[]> => {
+  const permissionListElementId = await awu(await elementSource.list())
```
Comment (alonstern) [inline packages/jira-adapter/src/change_validators/permission_type.ts:29]: I don't think you need to run list here, you already know the full elem id
  `new ElemID('jira', PERMISSIONS, 'instance', ElemID.CONFIG_NAME)`

File: packages/jira-adapter/src/change_validators/permission_type.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    .find(id => id.typeName === PERMISSIONS && id.idType === 'instance')
+
+  if (!permissionListElementId) {
+    return []
+  }
+  return Object.values(
+    (
+    await elementSource.get(permissionListElementId)
+    ).value.additionalProperties as { key: string }[]
```
Comment (alonstern) [inline packages/jira-adapter/src/change_validators/permission_type.ts:38]: nit:
```suggestion
    (await elementSource.get(permissionListElementId))
      .value.additionalProperties as { key: string }[]
```

Comment (alonstern) [inline packages/jira-adapter/src/change_validators/permission_type.ts:38]: can you avoid the casting by checking the type in runtime?

**Review by alonstern — COMMENTED** [review]
👌👌

File: packages/jira-adapter/src/change_validators/permission_type.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+const getInvalidPermissionErrorMessage = (
+  permissionScheme: InstanceElement,
+  allowedPermissions: string[],
+): string => {
+  const invalidPermissionTypes = Array.from(new Set(permissionScheme.value.permissions
+    .filter((permission: { permission: string }) =>
+      !allowedPermissions.includes(permission.permission))
+    .map((permission: { permission: string }) => permission.permission)))
+  return `The permissions ${invalidPermissionTypes.join(', ')} in ${permissionScheme.elemID.getFullName()} does not exist in the current environment and will be excluded during deployment`
```
[the remark this answers is not in the document]
Comment (guyhamo) [reply packages/jira-adapter/src/change_validators/permission_type.ts:58]: im going to trust you on this one.

**Review by alonstern — APPROVED** [review]

**Pull request merged**